src/sasl/cyrsasl_scram.erl

```diff
@@ -24,32 +25,40 @@ mech_new(LServer, Creds, #{sha := Sha,
                            auth_mech := AuthMech,
                            scram_plus := ScramPlus}) ->
     ChannelBinding = calculate_channel_binding(Socket, ScramPlus, Sha, AuthMech),
-    Fun = fun(Username, St0) ->
-                  JID = jid:make_bare(Username, LServer),
-                  HostType = mongoose_credentials:host_type(Creds),
-                  case get_scram_attributes(HostType, JID, Sha) of
-                      {AuthModule, {StoredKey, ServerKey, Salt, ItCount}} ->
-                          Creds1 = fast_scram:mech_get(creds, St0, Creds),
-                          R = [{username, Username}, {auth_module, AuthModule}],
-                          Creds2 = mongoose_credentials:extend(Creds1, R),
-                          St1 = fast_scram:mech_set(creds, Creds2, St0),
-                          ExtraConfig = #{it_count => ItCount, salt => Salt,
-                                          auth_data => #{stored_key => StoredKey,
-                                                         server_key => ServerKey}},
-                          {St1, ExtraConfig};
-                      {error, Reason, User} ->
-                          {error, {Reason, User}}
-                  end
-          end,
     {ok, St0} = fast_scram:mech_new(
                         #{entity => server,
                           hash_method => Sha,
-                          retrieve_mechanism => Fun,
+                          retrieve_mechanism => retrieve_mechanism_fun(LServer, Creds, Sha),
                           nonce_size => ?NONCE_LENGTH,
                           channel_binding => ChannelBinding}),
     St1 = fast_scram:mech_set(creds, Creds, St0),
     {ok, St1}.
 
+retrieve_mechanism_fun(LServer, Creds, Sha) ->
+    fun(Username, St0) ->
+            case jid:make_bare(Username, LServer) of
+                error -> {error, Username};
```

Comment:
So the error reason is the `Username`? Just double-checking: is this intended?

Reply:
Yes, `LServer` to be precise was already stringprepped by the time we arrived here. `stringprep:nodeprep/1` returns error if the jid has a leading whitespace, and the error was that we were passing this error to `ejabberd_auth:get_passterm_with_authmodule/2`, which expects an actual JID.

Reply:
Ok, and the scram library expects a {error, term()} tuple, this will be just given back to MIM as is and will be logged here: https://github.com/esl/MongooseIM/blob/b39a24905717119362a2bb8fb5c6143d6f4400d9/src/sasl/cyrsasl_scram.erl#L72